File: examples/voter/vote_reject.py
"""Cast ``voteReject`` on a disputed APEX job.

Usage:
    python vote_reject.py <jobId>

Performs three pre-flight checks before sending any transaction:
1. Caller is a whitelisted voter.
2. The job has actually been disputed.
3. The caller hasn't already voted.
"""

from __future__ import annotations

import os
import sys
from pathlib import Path

from dotenv import load_dotenv

from bnbagent.apex import APEXClient
from bnbagent.wallets import EVMWalletProvider

ROOT = Path(__file__).resolve().parent


def main() -> int:
    if len(sys.argv) != 2:
        print("Usage: python vote_reject.py <jobId>", file=sys.stderr)
        return 2

    try:
        job_id = int(sys.argv[1])
    except ValueError:
        print(f"jobId must be an integer, got {sys.argv[1]!r}", file=sys.stderr)
        return 2

    load_dotenv(ROOT / ".env")
    pk = os.environ.get("VOTER_PRIVATE_KEY")
    if not pk:
        print("VOTER_PRIVATE_KEY is required", file=sys.stderr)
        return 2

    network = os.environ.get("NETWORK", "bsc-testnet")
    wallet = EVMWalletProvider(password="example", private_key=pk, persist=False)
    apex = APEXClient(wallet, network=network)
    voter = apex.address
    assert voter is not None

    if not apex.policy.is_voter(voter):
        print(f"{voter} is NOT a whitelisted voter on {apex.policy.address}", file=sys.stderr)
        return 1
    if not apex.policy.disputed(job_id):
        print(f"jobId={job_id} has not been disputed yet; voteReject would revert", file=sys.stderr)
        return 1
    if apex.policy.has_voted(job_id, voter):
        print(f"{voter} already voted on jobId={job_id}", file=sys.stderr)
        return 0

    quorum = apex.policy.vote_quorum()
    current = apex.policy.reject_votes(job_id)
    print(f"[voter] casting voteReject on jobId={job_id} ({current}/{quorum} votes)")

    res = apex.vote_reject(job_id)
    print(f"[voter] tx: {res.get('tx_hash')}")

    new_total = apex.policy.reject_votes(job_id)
    if new_total >= quorum:
        print(f"[voter] quorum reached ({new_total}/{quorum}); any settler can now call router.settle({job_id})")
    else:
        print(f"[voter] current reject votes: {new_total}/{quorum} — still below quorum")
    return 0


if __name__ == "__main__":
    raise SystemExit(main())

File: typescript/examples/voter/voteReject.ts
/**
 * Cast `voteReject` on a disputed ERC-8183 job.
 *
 * Usage:
 *     pnpm -C typescript exec tsx examples/voter/voteReject.ts <jobId>
 *
 * Performs three pre-flight checks before sending any transaction:
 * 1. Caller is a whitelisted voter.
 * 2. The job has actually been disputed.
 * 3. The caller hasn't already voted.
 *
 * Port of `python/examples/voter/vote_reject.py`. Hits live testnet — run
 * manually, not in CI.
 */

import { dirname } from "node:path";
import { fileURLToPath } from "node:url";
import { ERC8183Client, EVMWalletProvider, loadEnv } from "../../src/index.js";

const ROOT = dirname(fileURLToPath(import.meta.url));

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

async function main(): Promise<number> {
  const jobIdArg = process.argv[2];
  if (!jobIdArg) {
    console.error("Usage: voteReject.ts <jobId>");
    return 2;
  }
  if (!/^\d+$/.test(jobIdArg)) {
    console.error(`jobId must be an integer, got ${JSON.stringify(jobIdArg)}`);
    return 2;
  }
  const jobId = BigInt(jobIdArg);

  loadEnv(ROOT);
  const pk = process.env.VOTER_PRIVATE_KEY;
  if (!pk) {
    console.error("VOTER_PRIVATE_KEY is required");
    return 2;
  }

  const network = process.env.NETWORK ?? "bsc-testnet";
  const wallet = new EVMWalletProvider({
    password: "example",
    privateKey: pk,
    persist: false,
  });
  const client = await ERC8183Client.create({
    walletProvider: wallet,
    network,
  });
  const voter = client.address;
  if (!voter) {
    throw new Error("voter address unavailable");
  }

  if (!(await client.policy.isVoter(voter))) {
    console.error(
      `${voter} is NOT a whitelisted voter on ${client.policy.address}`,
    );
    return 1;
  }
  if (!(await client.policy.disputed(jobId))) {
    console.error(
      `jobId=${jobId} has not been disputed yet; voteReject would revert`,
    );
    return 1;
  }
  if (await client.policy.hasVoted(jobId, voter)) {
    console.log(`${voter} already voted on jobId=${jobId}`);
    return 0;
  }

  const quorum = await client.policy.voteQuorum();
  const current = await client.policy.rejectVotes(jobId);
  console.log(
    `[voter] casting voteReject on jobId=${jobId} (${current}/${quorum} votes)`,
  );

  const res = await client.voteReject(jobId);
  console.log(`[voter] tx: ${res.transactionHash}`);

  // The RPC node may briefly serve a stale block — retry until the vote shows up.
  let newTotal = current;
  for (let i = 0; i < 8; i++) {
    newTotal = await client.policy.rejectVotes(jobId);
    if (newTotal > current) {
      break;
    }
    await sleep(2_000);
  }

  if (newTotal >= quorum) {
    console.log(
      `[voter] quorum reached (${newTotal}/${quorum}); any settler can now call router.settle(${jobId})`,
    );
  } else {
    console.log(
      `[voter] current reject votes: ${newTotal}/${quorum} — still below quorum`,
    );
  }
  return 0;
}

main()
  .then((code) => process.exit(code))
  .catch((error) => {
    console.error(error);
    process.exit(1);
  });
